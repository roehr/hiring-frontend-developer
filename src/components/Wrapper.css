.testimonial-container{
    display: flex;
    flex-direction: column;
    margin:32px;
    box-shadow: 0px 4px 42px 0px #4F72CD26;
}