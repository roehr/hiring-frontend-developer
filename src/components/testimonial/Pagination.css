
.pagination{
    display:flex;
    justify-content: space-between;
}

.pages{
    display:flex;
    justify-content: space-between;
}
.pagination-button{
    padding:5px;
    margin: 5px;
    border: 1px solid var(--button-border-color);
}
