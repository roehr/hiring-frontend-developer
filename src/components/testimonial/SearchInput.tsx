import React from "react";

interface SearchInputProps{
    onQueryInput: (query:string) => void
}

const SearchInput = (props: SearchInputProps) => {
    const [query, setQuery] = React.useState("")

    React.useEffect(()=>{
        const timeOutId = setTimeout(() => props.onQueryInput(query), 500);
        return () => clearTimeout(timeOutId);
    }, [query])

    return <div className="searchFilter">
        <input type="text" placeholder="Filter by Exercise Title" id="exercisefilter" onChange={event => setQuery(event.target.value)}/>
    </div>
}

export {SearchInput}