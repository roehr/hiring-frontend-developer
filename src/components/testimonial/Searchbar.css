.searchbar{
    min-height: 80px;
    align-items: center;
}