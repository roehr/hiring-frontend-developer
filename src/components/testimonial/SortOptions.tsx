import React from "react";

interface SortOptionsProps{
    onSelect: (order: string) => void
}
const SortOptions = (props: SortOptionsProps) => {
    return <div className="searchOptions">
        <select data-testid= "sort-options" name="order" id="order" onChange={(event)=>props.onSelect(event.target.value)}>
            <option value="newest_first">Sort by Most Recent</option>
            <option data-testid= "sort-options-oldest" value="oldest_first">Sort by Oldest First</option>
        </select>
    </div>
}

export {SortOptions}