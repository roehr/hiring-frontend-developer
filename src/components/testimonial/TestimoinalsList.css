
.testimonial-row{
    display: grid;
    grid-template-columns: 64px 64px 1fr 2fr 100px 60px;
    border-bottom: 1px solid #EAECF3;
    min-height: 64px;
    color: var(--label-default);
    text-decoration: none;
    align-items: center;
}
.testimonial-row--active{
  background: var(--hover-background)
}

.mentor-image img{
    border-radius: 50%;
}
@keyframes spin {
    0% {transform: rotate(0deg)}
    100% {transform: rotate(360deg)}
}

.testimonial-time{
    text-align: end;
}