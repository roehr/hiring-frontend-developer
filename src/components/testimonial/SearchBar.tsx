import React from "react";
import {TrackDropDown} from "./TrackDropDown";
import {SearchInput} from "./SearchInput";
import {SortOptions} from "./SortOptions";
import {EnhancedTrack, TrackData} from "../../tracks.model";
import {enhanceMappingForSearch} from "../../utils/enhanceTracks";
import "./Searchbar.css"
interface SearchBarProps{
    counts?: Object;
    trackList: TrackData[];
    totalCount: number;
    updateSlug: (slug:string) => void
    updateSearch: (exercise:string) => void
    updateOrder: (order: string) => void
}
const SearchBar = (props: SearchBarProps) => {
    let searchDropdown:EnhancedTrack[] = []
    if(props.counts){
       searchDropdown = enhanceMappingForSearch(props.counts, props.trackList, props.totalCount)
    }
    return(
        <div className="searchbar">
            <div className="tracksearchselection">
                <TrackDropDown enhancedData={searchDropdown}  updateData={props.updateSlug}/>
                <SearchInput  onQueryInput={props.updateSearch}/>
            </div>
            <div className="sortoption">
                <SortOptions onSelect={props.updateOrder} />
            </div>
        </div>
    )
}
export {SearchBar}