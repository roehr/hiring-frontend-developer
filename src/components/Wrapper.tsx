import {SearchBar} from "./SearchBar";
import {TestimonialsList} from "./TestimonialsList";
import React from "react";
import {Headline} from "./Headline";
import {Header} from "./Header";
import {Testimonials, TestimonialsRoot, Track} from "../testimonial.model";
import {getAllTracks, getTestimonials} from "../service/fetchData";
import {PaginationArea} from "./PaginationArea";
import {AllTracks, TrackData} from "../tracks.model";
import {enhanceMappingForSearch} from "../utils/enhanceTracks";

const Wrapper = () => {
    const [loading, isLoading] = React.useState(true);
    const [data, setdata] = React.useState(null as unknown as Testimonials)
    const [trackData, settrackData] = React.useState([] as TrackData[])
    const [order, setOrder] = React.useState("newest_first")
    const [page, setPage] = React.useState(1)
    const [enhancedTrackData, setEnhancedTrackData] = React.useState([])
    React.useEffect(() => {
        getAllTracks().then((data) => {
                settrackData(data.tracks)
            }
        ).catch()
    }, [])
    React.useEffect(() => {
        isLoading(true)
        getTestimonials(page, order).then(
            (data) => {
                setdata(data.testimonials);
                isLoading(false);
            }
        ).catch()
    }, [order, page])

    const onPaginationClicked = (newPage: number) => {
        setPage(newPage)
    }

    return <div id="exercism-testimonial">
        <Header/>
        <div className="mentoring-container">
            <Headline/>
            <SearchBar  counts={data&&data.track_counts? data.track_counts : undefined} trackList={trackData}/>

            {!loading && data.pagination && <PaginationArea pagination={data.pagination} onPageClicked={onPaginationClicked}/>}
        </div>
    </div>
}

export {Wrapper}