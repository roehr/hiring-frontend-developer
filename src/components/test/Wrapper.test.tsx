import {fireEvent, screen, render, waitFor, getByTestId} from "@testing-library/react";
import React from "react";
import {Wrapper} from "../Wrapper";
import {AllTracks} from "../../tracks.model";
import {TestimonialsRoot} from "../../testimonial.model";
import * as service from "../../service/fetchData";
import {expect} from "@jest/globals";
import userEvent from "@testing-library/user-event";
describe('Wrapper test', () => {

    test('renders No data', async () => {
        const tracks = {tracks: []} as AllTracks
        const testimonials = {testimonials: {
                results: [],
                pagination: {current_page:0, total_pages:0, total_count:0},
                tracks:[],
                track_counts: {}
            }} as TestimonialsRoot
        jest.spyOn(service, "getAllTracks").mockResolvedValueOnce(tracks)
        jest.spyOn(service, "getTestimonials").mockResolvedValueOnce(testimonials)
        render(<Wrapper />)
        expect(await screen.findByText("You currently have no testimonials")).toBeTruthy()
        expect(screen).toMatchSnapshot();
    });

    const tracks = {tracks: [{
            slug:'c',
            title:'C',
            course: true,
            icon_url: "cicon",
            has_notifications: true,
            is_joined: true,
            is_new: false,
            last_touched_at: new Date(),
            links: {self: 'cself', concepts:'C Concepts', exercises: "C EXC"},
            num_completed_exercises: 0,
            num_concepts: 2,
            num_exercises: 32,
            num_learnt_concepts: 22,
            num_solutions: 0,
            tags:[],
            web_url: "curl"
        },
            {
                slug:'cpp',
                title:'CPP',
                course: true,
                icon_url: "cicon",
                has_notifications: true,
                is_joined: true,
                is_new: false,
                last_touched_at: new Date(),
                links: {self: 'cppself', concepts:'Cpp Concepts', exercises: "CPP EXC"},
                num_completed_exercises: 0,
                num_concepts: 2,
                num_exercises: 32,
                num_learnt_concepts: 22,
                num_solutions: 0,
                tags:[],
                web_url: "cppurl"
            }]} as AllTracks
    const testimonials = {testimonials: {
            results: [
                {
                    id: 1234,
                    track: {
                        slug:"cpp",
                        icon_url:"urlimg",
                        title: "CPP"
                    },
                    mentor: {
                        avatar_url:"av_url",
                        handle:"handle"
                    },
                    content: "Some Content",
                    exercise: {
                        slug:"c",
                        icon_url:"urlimg",
                        title: "some ExerciseTitle",
                    },
                    created_at: new Date("2021-05-08T19:55:15.000Z")
                },
                {
                    id: 1244,
                    track: {
                        slug:"c",
                        icon_url:"urlimg",
                        title: "C"
                    },
                    mentor: {
                        avatar_url:"av_url",
                        handle:"handle"
                    },
                    content: "Some more Content",
                    exercise: {
                        slug:"c",
                        icon_url:"urlimg",
                        title: "some ExerciseTitle",
                    },
                    created_at: new Date("2021-05-08T19:55:15.000Z")
                },
                {
                    id: 1254,
                    track: {
                        slug:"cpp",
                        icon_url:"urlimg",
                        title: "CPP"
                    },
                    mentor: {
                        avatar_url:"av_url",
                        handle:"handle"
                    },
                    content: "Some Other Content",
                    exercise: {
                        slug:"cpp",
                        icon_url:"urlimg",
                        title: "some ExerciseTitle",
                    },
                    created_at: new Date("2021-05-08T19:55:15.000Z")
                }
            ],
            pagination: {
                current_page:2,
                total_pages:2,
                total_count:23},
            tracks:[
                "c",
                "cpp"],
            track_counts: {
                "c": 61,
                "cpp": 9,
            }
        }} as TestimonialsRoot

        test('renders With data', async () => {
            jest.spyOn(service, "getAllTracks").mockResolvedValue(tracks)
            const testimonialSpy = jest.spyOn(service, "getTestimonials")
            testimonialSpy.mockResolvedValue(testimonials)
            render(<Wrapper/>)
            expect(screen.getByTestId('container-loading')).toBeTruthy()
            await waitFor(() => screen.getByTestId('container-loaded'))
            expect(testimonialSpy).toHaveBeenCalledWith(1, expect.anything(), undefined, undefined);
        });

            test('select Track', async () => {
                jest.spyOn(service, "getAllTracks").mockResolvedValue(tracks)
                const testimonialSpy = jest.spyOn(service, "getTestimonials")
                testimonialSpy.mockResolvedValue(testimonials)
                render(<Wrapper/>)
                await waitFor(() => screen.getByTestId('container-loaded'))
                expect(testimonialSpy).toHaveBeenCalledWith(1, expect.anything(), undefined, undefined);
                fireEvent.click(screen.getByTestId("track-options-button"))
                await waitFor(() => screen.getByTestId('track-option-c'))
                userEvent.click(screen.getByTestId('track-option-c'))
                await waitFor(() => {expect(testimonialSpy).toHaveBeenCalledWith(1, expect.anything(), "c", undefined)})
            });

            test('select Sort', async () => {
                jest.spyOn(service, "getAllTracks").mockResolvedValue(tracks)
                const testimonialSpy = jest.spyOn(service, "getTestimonials")
                testimonialSpy.mockResolvedValue(testimonials)
                render(<Wrapper/>)
                await waitFor(() => screen.getByTestId('container-loaded'))
                expect(testimonialSpy).toHaveBeenCalledWith(1, expect.anything(), undefined, undefined);
                fireEvent.change(screen.getByTestId('sort-options'), { target: { value: 0 } })
                await waitFor(() => {expect(testimonialSpy).toHaveBeenCalledWith(1, "newest_first", undefined, undefined)})
            });

    test('select Page', async () => {
        jest.spyOn(service, "getAllTracks").mockResolvedValue(tracks)
        const testimonialSpy = jest.spyOn(service, "getTestimonials")
        testimonialSpy.mockResolvedValue(testimonials)
        render(<Wrapper/>)
        await waitFor(() => screen.getByTestId('prev-button'))
        expect(testimonialSpy).toHaveBeenCalledWith(1, expect.anything(), undefined, undefined);
        fireEvent.click(screen.getByTestId('prev-button'))
        await waitFor(() => {expect(testimonialSpy).toHaveBeenCalledWith(1, "newest_first", undefined, undefined)})
    });

    test('search Exercise', async () => {
        jest.spyOn(service, "getAllTracks").mockResolvedValue(tracks)
        const testimonialSpy = jest.spyOn(service, "getTestimonials")
        testimonialSpy.mockResolvedValue(testimonials)
        render(<Wrapper/>)
        await waitFor(() => screen.getByTestId('prev-button'))
        expect(testimonialSpy).toHaveBeenCalledWith(1, expect.anything(), undefined, undefined);
        fireEvent.change(screen.getByPlaceholderText('Filter by Exercise Title'), {target: {value: test}})
        await waitFor(() => {expect(testimonialSpy).toHaveBeenCalledWith(1, "newest_first", undefined, expect.anything())})
    });

    test('renders With Errors', async () => {
        jest.spyOn(service, "getAllTracks").mockRejectedValue("error")
        const testimonialSpy = jest.spyOn(service, "getTestimonials")
        testimonialSpy.mockRejectedValue("error")
        const underTest = render(<Wrapper/>)
        expect(underTest).toMatchSnapshot()
    });



});