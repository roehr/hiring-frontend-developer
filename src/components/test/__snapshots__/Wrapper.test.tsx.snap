// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wrapper test renders No data 1`] = `
Object {
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;

exports[`Wrapper test renders With Errors 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        id="testimonial"
      >
        <header>
          <div
            class="header-dummy"
          >
            <div
              class="logo"
            >
               Exercism Logo
            </div>
            <div
              class="dashboard"
            >
              Dashboard
            </div>
          </div>
        </header>
        <div
          class="hero"
        >
          <div
            class="hero-icon"
          >
            <img
              src="[object Object]"
            />
          </div>
          <div
            class="hero-text"
          >
            <h2>
              Testimonials I've left
            </h2>
          </div>
          <div
            class="hero-line"
          >
            <img
              src="[object Object]"
            />
          </div>
        </div>
        <div
          class="testimonial-container"
          data-testid="container-loading"
        >
          <div
            class="searchbar"
          >
            <div
              class="tracksearchselection"
            >
              <div
                class="trackdropdown"
              >
                <button
                  aria-haspopup="true"
                  class="trackbutton"
                  data-testid="track-options-button"
                >
                  <img
                    class="track-icon"
                    src="https://d24y9kuxp2d7l2.cloudfront.net/assets/icons/all-tracks-5723af830c1df04c577e2aefdc8003f62c4270ec.svg"
                  />
                  <img
                    class="chevron"
                    src="https://d24y9kuxp2d7l2.cloudfront.net/assets/icons/chevron-down-5ae28e42ee217bae38f4eb1c119cafd0301dd5f6.svg"
                  />
                </button>
              </div>
              <div
                class="searchFilter"
              >
                <input
                  id="exercisefilter"
                  placeholder="Filter by Exercise Title"
                  type="text"
                />
              </div>
            </div>
            <div
              class="sortoption"
            >
              <div
                class="searchOptions"
              >
                <select
                  data-testid="sort-options"
                  id="order"
                  name="order"
                >
                  <option
                    value="newest_first"
                  >
                    Sort by Most Recent
                  </option>
                  <option
                    data-testid="sort-options-oldest"
                    value="oldest_first"
                  >
                    Sort by Oldest First
                  </option>
                </select>
              </div>
            </div>
          </div>
          <div
            class="testimonial-zone"
          >
            <div
              class="--fetching-overlay flex items-center"
            >
              <img
                alt="Loading data"
                class="icon c-icon--spin"
                src="https://d24y9kuxp2d7l2.cloudfront.net/assets/icons/spinner-0813469c6c8f037b2e3d8aa5e987173eec009c02.svg"
              />
            </div>
            <div
              class="testimonials"
            >
              <div
                class="no-testimonials"
              >
                <img
                  alt=""
                  class="icon"
                  role="presentation"
                  src="https://d24y9kuxp2d7l2.cloudfront.net/assets/icons/testimonials-15961f7587c64f2c588fb12b520cf2b87d029155.svg"
                />
                <h2>
                  You currently have no testimonials
                </h2>
                <p>
                  Testimonials are left by students on ending successful mentoring discussions.
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      id="testimonial"
    >
      <header>
        <div
          class="header-dummy"
        >
          <div
            class="logo"
          >
             Exercism Logo
          </div>
          <div
            class="dashboard"
          >
            Dashboard
          </div>
        </div>
      </header>
      <div
        class="hero"
      >
        <div
          class="hero-icon"
        >
          <img
            src="[object Object]"
          />
        </div>
        <div
          class="hero-text"
        >
          <h2>
            Testimonials I've left
          </h2>
        </div>
        <div
          class="hero-line"
        >
          <img
            src="[object Object]"
          />
        </div>
      </div>
      <div
        class="testimonial-container"
        data-testid="container-loading"
      >
        <div
          class="searchbar"
        >
          <div
            class="tracksearchselection"
          >
            <div
              class="trackdropdown"
            >
              <button
                aria-haspopup="true"
                class="trackbutton"
                data-testid="track-options-button"
              >
                <img
                  class="track-icon"
                  src="https://d24y9kuxp2d7l2.cloudfront.net/assets/icons/all-tracks-5723af830c1df04c577e2aefdc8003f62c4270ec.svg"
                />
                <img
                  class="chevron"
                  src="https://d24y9kuxp2d7l2.cloudfront.net/assets/icons/chevron-down-5ae28e42ee217bae38f4eb1c119cafd0301dd5f6.svg"
                />
              </button>
            </div>
            <div
              class="searchFilter"
            >
              <input
                id="exercisefilter"
                placeholder="Filter by Exercise Title"
                type="text"
              />
            </div>
          </div>
          <div
            class="sortoption"
          >
            <div
              class="searchOptions"
            >
              <select
                data-testid="sort-options"
                id="order"
                name="order"
              >
                <option
                  value="newest_first"
                >
                  Sort by Most Recent
                </option>
                <option
                  data-testid="sort-options-oldest"
                  value="oldest_first"
                >
                  Sort by Oldest First
                </option>
              </select>
            </div>
          </div>
        </div>
        <div
          class="testimonial-zone"
        >
          <div
            class="--fetching-overlay flex items-center"
          >
            <img
              alt="Loading data"
              class="icon c-icon--spin"
              src="https://d24y9kuxp2d7l2.cloudfront.net/assets/icons/spinner-0813469c6c8f037b2e3d8aa5e987173eec009c02.svg"
            />
          </div>
          <div
            class="testimonials"
          >
            <div
              class="no-testimonials"
            >
              <img
                alt=""
                class="icon"
                role="presentation"
                src="https://d24y9kuxp2d7l2.cloudfront.net/assets/icons/testimonials-15961f7587c64f2c588fb12b520cf2b87d029155.svg"
              />
              <h2>
                You currently have no testimonials
              </h2>
              <p>
                Testimonials are left by students on ending successful mentoring discussions.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
